{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirna_df = pd.read_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\fields\\SRR14432308\\mirna_res.csv\", header=None)\n",
    "columns = ['qaccver', 'qlen','saccver','slen','qstart','qend','sstart','send','qseq','sseq','evalue','bitscore','length','sstrand', 'gapopen', 'mismatch']\n",
    "mirna_df.columns = columns\n",
    "#mirna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 6981\n",
      "Unique miRNAs: 884\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mirna_df['qaccver'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mirna_df['saccver'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qaccver</th>\n",
       "      <th>qlen</th>\n",
       "      <th>saccver</th>\n",
       "      <th>slen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>sstrand</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>mismatch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR14432308.1024</td>\n",
       "      <td>74</td>\n",
       "      <td>hsa-miR-3195</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30.1</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR14432308.1267</td>\n",
       "      <td>84</td>\n",
       "      <td>hsa-miR-3195</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30.1</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR14432308.2037</td>\n",
       "      <td>74</td>\n",
       "      <td>hsa-miR-3195</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30.1</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR14432308.2042</td>\n",
       "      <td>54</td>\n",
       "      <td>hsa-miR-7704</td>\n",
       "      <td>19</td>\n",
       "      <td>38</td>\n",
       "      <td>50</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>GGTCGGCGGCGAC</td>\n",
       "      <td>GGTCGGCGGCGAC</td>\n",
       "      <td>0.029</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR14432308.2350</td>\n",
       "      <td>76</td>\n",
       "      <td>hsa-miR-6717-5p</td>\n",
       "      <td>22</td>\n",
       "      <td>39</td>\n",
       "      <td>56</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>GCGATGTGGGGACGGAGA</td>\n",
       "      <td>GCGATGTGGGGATGTAGA</td>\n",
       "      <td>0.045</td>\n",
       "      <td>24.7</td>\n",
       "      <td>18</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8819</th>\n",
       "      <td>SRR14432308.15176428</td>\n",
       "      <td>51</td>\n",
       "      <td>hsa-miR-6749-3p</td>\n",
       "      <td>21</td>\n",
       "      <td>33</td>\n",
       "      <td>45</td>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "      <td>CTGCCTGGCCCAG</td>\n",
       "      <td>CTGCCTGGCCCAG</td>\n",
       "      <td>0.027</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8820</th>\n",
       "      <td>SRR14432308.15178087</td>\n",
       "      <td>75</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>63</td>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>0.044</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8821</th>\n",
       "      <td>SRR14432308.15183391</td>\n",
       "      <td>152</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>137</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>GAGTTCTACAGTCCGA</td>\n",
       "      <td>GAGTTCTACAGTCAGA</td>\n",
       "      <td>0.086</td>\n",
       "      <td>25.6</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8822</th>\n",
       "      <td>SRR14432308.15188564</td>\n",
       "      <td>170</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>153</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>0.098</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8823</th>\n",
       "      <td>SRR14432308.15192174</td>\n",
       "      <td>60</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>0.033</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8824 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   qaccver  qlen          saccver  slen  qstart  qend  sstart  \\\n",
       "0         SRR14432308.1024    74     hsa-miR-3195    17      13    28       1   \n",
       "1         SRR14432308.1267    84     hsa-miR-3195    17      22    37       1   \n",
       "2         SRR14432308.2037    74     hsa-miR-3195    17      13    28       1   \n",
       "3         SRR14432308.2042    54     hsa-miR-7704    19      38    50       4   \n",
       "4         SRR14432308.2350    76  hsa-miR-6717-5p    22      39    56       3   \n",
       "...                    ...   ...              ...   ...     ...   ...     ...   \n",
       "8819  SRR14432308.15176428    51  hsa-miR-6749-3p    21      33    45       9   \n",
       "8820  SRR14432308.15178087    75     hsa-miR-3168    17      63    75       1   \n",
       "8821  SRR14432308.15183391   152     hsa-miR-3168    17     137   152       1   \n",
       "8822  SRR14432308.15188564   170     hsa-miR-3168    17     153   165       1   \n",
       "8823  SRR14432308.15192174    60     hsa-miR-3168    17       3    15       1   \n",
       "\n",
       "      send                qseq                sseq  evalue  bitscore  length  \\\n",
       "0       16    CGCGCCGGGCCCGGGT    CGCGCCGGGCCCGGGT   0.001      30.1      16   \n",
       "1       16    CGCGCCGGGCCCGGGT    CGCGCCGGGCCCGGGT   0.001      30.1      16   \n",
       "2       16    CGCGCCGGGCCCGGGT    CGCGCCGGGCCCGGGT   0.001      30.1      16   \n",
       "3       16       GGTCGGCGGCGAC       GGTCGGCGGCGAC   0.029      24.7      13   \n",
       "4       20  GCGATGTGGGGACGGAGA  GCGATGTGGGGATGTAGA   0.045      24.7      18   \n",
       "...    ...                 ...                 ...     ...       ...     ...   \n",
       "8819    21       CTGCCTGGCCCAG       CTGCCTGGCCCAG   0.027      24.7      13   \n",
       "8820    13       GAGTTCTACAGTC       GAGTTCTACAGTC   0.044      24.7      13   \n",
       "8821    16    GAGTTCTACAGTCCGA    GAGTTCTACAGTCAGA   0.086      25.6      16   \n",
       "8822    13       GAGTTCTACAGTC       GAGTTCTACAGTC   0.098      24.7      13   \n",
       "8823    13       GAGTTCTACAGTC       GAGTTCTACAGTC   0.033      24.7      13   \n",
       "\n",
       "     sstrand  gapopen  mismatch  \n",
       "0       plus        0         0  \n",
       "1       plus        0         0  \n",
       "2       plus        0         0  \n",
       "3       plus        0         0  \n",
       "4       plus        0         2  \n",
       "...      ...      ...       ...  \n",
       "8819    plus        0         0  \n",
       "8820    plus        0         0  \n",
       "8821    plus        0         1  \n",
       "8822    plus        0         0  \n",
       "8823    plus        0         0  \n",
       "\n",
       "[8824 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#removing antisense strands\n",
    "mirna_df.drop(mirna_df[mirna_df['sstrand'] == 'minus'].index, inplace=True)\n",
    "mirna_df.reset_index(drop=True, inplace=True)\n",
    "mirna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 6981\n",
      "Unique miRNAs: 884\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mirna_df['qaccver'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mirna_df['saccver'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qaccver</th>\n",
       "      <th>qlen</th>\n",
       "      <th>saccver</th>\n",
       "      <th>slen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>qseq</th>\n",
       "      <th>sseq</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>length</th>\n",
       "      <th>sstrand</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>mismatch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR14432308.1024</td>\n",
       "      <td>74</td>\n",
       "      <td>hsa-miR-3195</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30.1</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR14432308.1267</td>\n",
       "      <td>84</td>\n",
       "      <td>hsa-miR-3195</td>\n",
       "      <td>17</td>\n",
       "      <td>22</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30.1</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR14432308.2037</td>\n",
       "      <td>74</td>\n",
       "      <td>hsa-miR-3195</td>\n",
       "      <td>17</td>\n",
       "      <td>13</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>CGCGCCGGGCCCGGGT</td>\n",
       "      <td>0.001</td>\n",
       "      <td>30.1</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR14432308.2042</td>\n",
       "      <td>54</td>\n",
       "      <td>hsa-miR-7704</td>\n",
       "      <td>19</td>\n",
       "      <td>38</td>\n",
       "      <td>50</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>GGTCGGCGGCGAC</td>\n",
       "      <td>GGTCGGCGGCGAC</td>\n",
       "      <td>0.029</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR14432308.2350</td>\n",
       "      <td>76</td>\n",
       "      <td>hsa-miR-6717-5p</td>\n",
       "      <td>22</td>\n",
       "      <td>39</td>\n",
       "      <td>56</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>GCGATGTGGGGACGGAGA</td>\n",
       "      <td>GCGATGTGGGGATGTAGA</td>\n",
       "      <td>0.045</td>\n",
       "      <td>24.7</td>\n",
       "      <td>18</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8748</th>\n",
       "      <td>SRR14432308.15176428</td>\n",
       "      <td>51</td>\n",
       "      <td>hsa-miR-6749-3p</td>\n",
       "      <td>21</td>\n",
       "      <td>33</td>\n",
       "      <td>45</td>\n",
       "      <td>9</td>\n",
       "      <td>21</td>\n",
       "      <td>CTGCCTGGCCCAG</td>\n",
       "      <td>CTGCCTGGCCCAG</td>\n",
       "      <td>0.027</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8749</th>\n",
       "      <td>SRR14432308.15178087</td>\n",
       "      <td>75</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>63</td>\n",
       "      <td>75</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>0.044</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8750</th>\n",
       "      <td>SRR14432308.15183391</td>\n",
       "      <td>152</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>137</td>\n",
       "      <td>152</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>GAGTTCTACAGTCCGA</td>\n",
       "      <td>GAGTTCTACAGTCAGA</td>\n",
       "      <td>0.086</td>\n",
       "      <td>25.6</td>\n",
       "      <td>16</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8751</th>\n",
       "      <td>SRR14432308.15188564</td>\n",
       "      <td>170</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>153</td>\n",
       "      <td>165</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>0.098</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8752</th>\n",
       "      <td>SRR14432308.15192174</td>\n",
       "      <td>60</td>\n",
       "      <td>hsa-miR-3168</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>0.033</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>plus</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8753 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   qaccver  qlen          saccver  slen  qstart  qend  sstart  \\\n",
       "0         SRR14432308.1024    74     hsa-miR-3195    17      13    28       1   \n",
       "1         SRR14432308.1267    84     hsa-miR-3195    17      22    37       1   \n",
       "2         SRR14432308.2037    74     hsa-miR-3195    17      13    28       1   \n",
       "3         SRR14432308.2042    54     hsa-miR-7704    19      38    50       4   \n",
       "4         SRR14432308.2350    76  hsa-miR-6717-5p    22      39    56       3   \n",
       "...                    ...   ...              ...   ...     ...   ...     ...   \n",
       "8748  SRR14432308.15176428    51  hsa-miR-6749-3p    21      33    45       9   \n",
       "8749  SRR14432308.15178087    75     hsa-miR-3168    17      63    75       1   \n",
       "8750  SRR14432308.15183391   152     hsa-miR-3168    17     137   152       1   \n",
       "8751  SRR14432308.15188564   170     hsa-miR-3168    17     153   165       1   \n",
       "8752  SRR14432308.15192174    60     hsa-miR-3168    17       3    15       1   \n",
       "\n",
       "      send                qseq                sseq  evalue  bitscore  length  \\\n",
       "0       16    CGCGCCGGGCCCGGGT    CGCGCCGGGCCCGGGT   0.001      30.1      16   \n",
       "1       16    CGCGCCGGGCCCGGGT    CGCGCCGGGCCCGGGT   0.001      30.1      16   \n",
       "2       16    CGCGCCGGGCCCGGGT    CGCGCCGGGCCCGGGT   0.001      30.1      16   \n",
       "3       16       GGTCGGCGGCGAC       GGTCGGCGGCGAC   0.029      24.7      13   \n",
       "4       20  GCGATGTGGGGACGGAGA  GCGATGTGGGGATGTAGA   0.045      24.7      18   \n",
       "...    ...                 ...                 ...     ...       ...     ...   \n",
       "8748    21       CTGCCTGGCCCAG       CTGCCTGGCCCAG   0.027      24.7      13   \n",
       "8749    13       GAGTTCTACAGTC       GAGTTCTACAGTC   0.044      24.7      13   \n",
       "8750    16    GAGTTCTACAGTCCGA    GAGTTCTACAGTCAGA   0.086      25.6      16   \n",
       "8751    13       GAGTTCTACAGTC       GAGTTCTACAGTC   0.098      24.7      13   \n",
       "8752    13       GAGTTCTACAGTC       GAGTTCTACAGTC   0.033      24.7      13   \n",
       "\n",
       "     sstrand  gapopen  mismatch  \n",
       "0       plus        0         0  \n",
       "1       plus        0         0  \n",
       "2       plus        0         0  \n",
       "3       plus        0         0  \n",
       "4       plus        0         2  \n",
       "...      ...      ...       ...  \n",
       "8748    plus        0         0  \n",
       "8749    plus        0         0  \n",
       "8750    plus        0         1  \n",
       "8751    plus        0         0  \n",
       "8752    plus        0         0  \n",
       "\n",
       "[8753 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#removing alignment loops\n",
    "mirna_df.drop(mirna_df[mirna_df['gapopen'] > 0].index, inplace=True)\n",
    "mirna_df.reset_index(drop=True, inplace=True)\n",
    "mirna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 6933\n",
      "Unique miRNAs: 877\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mirna_df['qaccver'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mirna_df['saccver'].nunique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# preprocessing mRNA results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pandas as pd\n",
    "mrna_df = pd.read_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\fields\\SRR14432308\\mrna_res.csv\", header=None)\n",
    "columns = ['qaccver', 'qlen','saccver','slen','qstart','qend','sstart','send','qseq','sseq','evalue','bitscore','length','sstrand', 'gapopen']\n",
    "mrna_df.columns = columns\n",
    "#mrna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 16638\n",
      "Unique mRNAs: 44891\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_df['saccver'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#removing antisense strands\n",
    "mrna_df.drop(mrna_df[mrna_df['sstrand'] == 'minus'].index, inplace=True)\n",
    "mrna_df.reset_index(drop=True, inplace=True)\n",
    "#mrna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 16638\n",
      "Unique mRNAs: 44891\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_df['saccver'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#removing alignment loops\n",
    "mrna_df.drop(mrna_df[mrna_df['gapopen'] > 0].index, inplace=True)\n",
    "mrna_df.reset_index(drop=True, inplace=True)\n",
    "#mrna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 11646\n",
      "Unique mRNAs: 37044\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_df['saccver'].nunique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting Chimeric Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads in miRNA: 6933\n",
      "Unique Reads in mRNA: 11646\n"
     ]
    }
   ],
   "source": [
    "reads_from_miRNA = list(mirna_df['qaccver'].unique())\n",
    "reads_from_mRNA = list(mrna_df['qaccver'].unique())\n",
    "print('Unique Reads in miRNA: {}'.format(len(reads_from_miRNA)))\n",
    "print('Unique Reads in mRNA: {}'.format(len(reads_from_mRNA)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1508\n"
     ]
    }
   ],
   "source": [
    "#extracting common reads from both mRNA and miRNA which will be common reads.\n",
    "chimeric_reads = list(set(reads_from_miRNA) & set(reads_from_mRNA))\n",
    "print(len(chimeric_reads))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "chimeric_mrna_df = mrna_df[mrna_df['qaccver'].isin(chimeric_reads)]\n",
    "chimeric_mrna_df.reset_index(drop=True, inplace=True)\n",
    "chimeric_mirna_df = mirna_df[mirna_df['qaccver'].isin(chimeric_reads)]\n",
    "chimeric_mirna_df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\16892\\AppData\\Local\\Temp\\ipykernel_14908\\77221104.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  chimeric_mirna_df.drop(columns=['sstrand', 'gapopen'], inplace=True)\n",
      "C:\\Users\\16892\\AppData\\Local\\Temp\\ipykernel_14908\\77221104.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  chimeric_mrna_df.drop(columns=['sstrand', 'gapopen'], inplace=True)\n"
     ]
    }
   ],
   "source": [
    "#dropping unnecessary columns\n",
    "chimeric_mirna_df.drop(columns=['sstrand', 'gapopen'], inplace=True)\n",
    "chimeric_mrna_df.drop(columns=['sstrand', 'gapopen'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df = pd.merge(chimeric_mrna_df,chimeric_mirna_df, on='qaccver', how= 'inner', suffixes=('_mRNA', '_miRNA'))\n",
    "#mrna_mirna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 1508\n",
      "Unique mRNAs: 13541\n",
      "Unique miRNAs: 516\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_mirna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_mirna_df['saccver_mRNA'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mrna_mirna_df['saccver_miRNA'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df.to_csv(r'C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\eclip\\SRR18281063\\R1\\mrna_mirna.csv', index= False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#keeping the reads with significant miRNA portion (the same miRNA portion occured in at least 10 chimeric reads)\n",
    "filtered_df = mrna_mirna_df.groupby('qseq_miRNA').filter(lambda x: x['qaccver'].nunique() >= 10)\n",
    "filtered_df.reset_index(drop=True, inplace=True)\n",
    "mrna_mirna_df =filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 765\n",
      "Unique mRNAs: 4010\n",
      "Unique miRNAs: 6\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_mirna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_mirna_df['saccver_mRNA'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mrna_mirna_df['saccver_miRNA'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df.to_csv(r'C:\\Users\\16892\\Downloads\\blast_tutorial\\eclip\\SRR18281061\\mrna_mirna_10.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## marking ovelapping mRNA-miRNA portion sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df['gap'] = 0\n",
    "#mrna_mirna_df['gap'] = mrna_mirna_df['gap'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_condition = (((mrna_mirna_df['qstart_mRNA'] >= mrna_mirna_df['qstart_miRNA']) & (mrna_mirna_df['qstart_mRNA'] <= mrna_mirna_df['qend_miRNA'])) | \n",
    "                     ((mrna_mirna_df['qstart_miRNA'] >= mrna_mirna_df['qstart_mRNA']) & (mrna_mirna_df['qstart_miRNA'] <= mrna_mirna_df['qend_mRNA'])))\n",
    "\n",
    "mrna_mirna_df.loc[overlap_condition, 'gap'] = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "gap_condition_1 = ((mrna_mirna_df['qend_miRNA'] <= mrna_mirna_df['qstart_mRNA']) & (mrna_mirna_df['gap'] != -1))\n",
    "\n",
    "mrna_mirna_df.loc[gap_condition_1, 'gap'] = mrna_mirna_df['qstart_mRNA'] - mrna_mirna_df['qend_miRNA'] - 1\n",
    "\n",
    "gap_condition_2 = ((mrna_mirna_df['qend_mRNA'] <= mrna_mirna_df['qstart_miRNA']) & (mrna_mirna_df['gap'] != -1))\n",
    "\n",
    "mrna_mirna_df.loc[gap_condition_2, 'gap'] = mrna_mirna_df['qstart_miRNA'] - mrna_mirna_df['qend_mRNA'] - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df = mrna_mirna_df[mrna_mirna_df['gap'] <= 4] #allowing maximum 4nt gap\n",
    "mrna_mirna_df.reset_index(drop=True, inplace=True)\n",
    "#mrna_mirna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\16892\\AppData\\Local\\Temp\\ipykernel_14908\\369035784.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mrna_mirna_df['overlap'] = np.maximum(overlap1, overlap2)\n"
     ]
    }
   ],
   "source": [
    "start1 = mrna_mirna_df['qstart_mRNA']\n",
    "end1 = mrna_mirna_df['qend_mRNA']\n",
    "start2 = mrna_mirna_df['qstart_miRNA']\n",
    "end2 = mrna_mirna_df['qend_miRNA']\n",
    "\n",
    "# Overlap calculations\n",
    "overlap1 = np.where(\n",
    "    (start2 >= start1) & (start2 <= end1),\n",
    "    np.minimum(end1, end2) - start2 + 1,\n",
    "    0\n",
    ")\n",
    "\n",
    "overlap2 = np.where(\n",
    "    (start1 >= start2) & (start1 <= end2),\n",
    "    np.minimum(end1, end2) - start1 + 1,\n",
    "    0\n",
    ")\n",
    "\n",
    "# Combine the overlap calculations\n",
    "mrna_mirna_df['overlap'] = np.maximum(overlap1, overlap2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df = mrna_mirna_df[mrna_mirna_df['overlap'] <= 4] #allowing maximum 4 nucleotide overlap\n",
    "mrna_mirna_df.reset_index(drop=True, inplace=True)\n",
    "#mrna_mirna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 52\n",
      "Unique mRNAs: 605\n",
      "Unique miRNAs: 5\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_mirna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_mirna_df['saccver_mRNA'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mrna_mirna_df['saccver_miRNA'].nunique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keeping most significant miRNA-mRNA pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_most_significant_pair(df):\n",
    "    df_sorted = df.sort_values(by = ['evalue_miRNA', 'evalue_mRNA', 'bitscore_miRNA', 'bitscore_mRNA'], ascending = [True, True, False, False])\n",
    "    return df_sorted.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\16892\\AppData\\Local\\Temp\\ipykernel_14908\\3558051312.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  significant_df = mrna_mirna_df.groupby('qaccver').apply(get_most_significant_pair).reset_index(drop=True)\n"
     ]
    }
   ],
   "source": [
    "significant_df = mrna_mirna_df.groupby('qaccver').apply(get_most_significant_pair).reset_index(drop=True)\n",
    "mrna_mirna_df = significant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Reads: 52\n",
      "Unique mRNAs: 47\n",
      "Unique miRNAs: 4\n"
     ]
    }
   ],
   "source": [
    "print('Unique Reads: {}'.format(mrna_mirna_df['qaccver'].nunique()))\n",
    "print('Unique mRNAs: {}'.format(mrna_mirna_df['saccver_mRNA'].nunique()))\n",
    "print('Unique miRNAs: {}'.format(mrna_mirna_df['saccver_miRNA'].nunique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df.to_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\fields\\SRR14432308\\final_mrna_mirna.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Categorizing miRNA-mRNA and isomiR-mRNA interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df = pd.read_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\fields\\SRR13415091\\final_mrna_mirna.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(fasta_file):\n",
    "    \"\"\"\n",
    "    Reads a FASTA file and returns a dictionary of sequences.\n",
    "\n",
    "    Args:\n",
    "    fasta_file (str): Path to the FASTA file.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary with sequence identifiers as keys and sequences as values.\n",
    "    \"\"\"\n",
    "    sequences = {}\n",
    "    with open(fasta_file, 'r') as file:\n",
    "        sequence_id = \"\"\n",
    "        sequence = \"\"\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if sequence_id:\n",
    "                    sequences[sequence_id] = sequence\n",
    "                    sequence = \"\"\n",
    "                sequence_id = line[1:]  # Remove '>' character\n",
    "                sequence_id = sequence_id.split()[0] #keeping only miRNA name.\n",
    "            else:\n",
    "                sequence += line\n",
    "                sequence = sequence.replace(\"U\", \"T\") #replacing U with T since blast replaces U with T\n",
    "                \n",
    "        if sequence_id:\n",
    "            sequences[sequence_id] = sequence\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences = read_fasta(r\"C:\\Users\\16892\\Downloads\\blast_tutorial\\queries\\mirbase.fasta\")\n",
    "ori_seq = read_fasta(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\datasets\\fields_data\\SRR14432308\\SRR14432308.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fastq_quality_to_scores(quality_string):\n",
    "    \"\"\"\n",
    "    Convert a FASTQ quality string to a list of quality scores.\n",
    "\n",
    "    Args:\n",
    "    quality_string (str): A string of ASCII characters representing quality scores.\n",
    "\n",
    "    Returns:\n",
    "    List[int]: A list of numerical quality scores.\n",
    "    \"\"\"\n",
    "    return [ord(char) - 33 for char in quality_string]\n",
    "\n",
    "\n",
    "import gzip\n",
    "\n",
    "def read_fastq(file_path):\n",
    "    sequences = {}\n",
    "    with gzip.open(file_path, 'rt') as file:\n",
    "        while True:\n",
    "            try:\n",
    "                seq_id = file.readline().strip()  # Sequence identifier\n",
    "                seq = file.readline().strip()     # Sequence\n",
    "                plus = file.readline().strip()    # Plus line\n",
    "                qual = file.readline().strip()    # Quality scores\n",
    "                \n",
    "                if not seq_id:\n",
    "                    break\n",
    "                \n",
    "                if seq_id.startswith('@') and plus.startswith('+'):\n",
    "                    # Strip the '@' from the sequence identifier\n",
    "                    seq_id = seq_id[1:].split()[0]\n",
    "                    sequences[seq_id] = fastq_quality_to_scores(qual)\n",
    "                else:\n",
    "                    raise ValueError(\"Incorrect FASTQ format\")\n",
    "            \n",
    "            except Exception as e:\n",
    "                print(f\"Error: {e}\")\n",
    "                break\n",
    "                \n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "quality_sequneces = read_fastq(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\datasets\\fields_data\\SRR14432308\\SRR14432308_unique.fastq.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#keeping only aligned chimeric miRNA portion's quality scores & the chimeric reads from final_mrna_mirna\n",
    "'''chim_qual_seq = {}\n",
    "for each in mrna_mirna_df['qaccver'].unique():\n",
    "    qual_seq = quality_sequneces[each]\n",
    "    start_ind = mrna_mirna_df.loc[mrna_mirna_df['qaccver'] == each, 'qstart_miRNA'].iat[0] - 1\n",
    "    end_ind = mrna_mirna_df.loc[mrna_mirna_df['qaccver'] == each, 'qend_miRNA'].iat[0] - 1\n",
    "    chim_qual_seq[each] = qual_seq[start_ind:end_ind+1]'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#keeping only aligned chimeric miRNA portion's quality scores & the chimeric reads from final_mrna_mirna\n",
    "chim_qual_seq = {}\n",
    "\n",
    "# Group the DataFrame by 'qaccver' and iterate over each group\n",
    "grouped_df = mrna_mirna_df.groupby('qaccver')\n",
    "for each, group in grouped_df:\n",
    "    qual_seq = quality_sequneces[each]\n",
    "    start_ind = group['qstart_miRNA'].iloc[0] - 1\n",
    "    end_ind = group['qend_miRNA'].iloc[0] - 1\n",
    "    chim_qual_seq[each] = qual_seq[start_ind:end_ind+1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keeping the rest of the sequence excluding mRNA portion in a special sequence column \n",
    "\n",
    "mrna_mirna_df[\"spec_seq\"] = \"\" \n",
    "for i in range(len(mrna_mirna_df)):\n",
    "    temp_seq = ori_seq[mrna_mirna_df['qaccver'][i]]\n",
    "    if mrna_mirna_df['qstart_miRNA'][i] >= mrna_mirna_df['qend_mRNA'][i]: #miRNA is on the right of mRNA\n",
    "        if mrna_mirna_df['gap'][i] == -1:\n",
    "            start_ind = mrna_mirna_df['qstart_miRNA'][i] - 1\n",
    "            end_ind = mrna_mirna_df['qlen_miRNA'][i]\n",
    "        else:\n",
    "            start_ind = mrna_mirna_df['qstart_miRNA'][i] - int(mrna_mirna_df['gap'][i]) - 1\n",
    "            end_ind = mrna_mirna_df['qlen_miRNA'][i]\n",
    "            \n",
    "    else: # miRNA is on the left of mRNA\n",
    "        if mrna_mirna_df['gap'][i] == -1:\n",
    "            start_ind = 0\n",
    "            end_ind = mrna_mirna_df['qend_miRNA'][i]\n",
    "        else:\n",
    "            start_ind = 0\n",
    "            end_ind = mrna_mirna_df['qend_miRNA'][i] + int(mrna_mirna_df['gap'][i])\n",
    "    #print(start_ind, end_ind)\n",
    "    final_seq = temp_seq[start_ind:end_ind]\n",
    "    mrna_mirna_df[\"spec_seq\"].iat[i] = final_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#checking the slicing of special sequence was right or not\n",
    "\n",
    "for i in range(len(mrna_mirna_df)):\n",
    "    if mrna_mirna_df['qseq_miRNA'][i] not in mrna_mirna_df['spec_seq'][i]:\n",
    "        print(mrna_mirna_df['qseq_miRNA'][i])\n",
    "        print(mrna_mirna_df['spec_seq'][i])\n",
    "        print('YES')\n",
    "        print(i)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_quality(qual_seq):\n",
    "    for i in range(len(qual_seq)):\n",
    "        if qual_seq[i] < 30:\n",
    "            return 1\n",
    "    return 0\n",
    "\n",
    "def get_qual_flag(seq_id, i):\n",
    "    qual_seq = quality_sequneces[seq_id]\n",
    "    qual_flag = 0\n",
    "    if mrna_mirna_df['qstart_miRNA'][i] >= mrna_mirna_df['qend_mRNA'][i]: #miRNA is on the right of mRNA\n",
    "        end_ind = mrna_mirna_df['qlen_miRNA'][i]\n",
    "        if mrna_mirna_df['gap'][i] <= 0:\n",
    "            qual_flag = check_quality(qual_seq[mrna_mirna_df['qend_miRNA'][i]-1:end_ind])\n",
    "        else:\n",
    "            start_ind = mrna_mirna_df['qstart_miRNA'][i] - int(mrna_mirna_df['gap'][i]) - 1\n",
    "            qual_flag = check_quality(qual_seq[start_ind:mrna_mirna_df['qstart_miRNA'][i]])\n",
    "\n",
    "            if qual_flag == 0: \n",
    "             qual_flag = check_quality(qual_seq[mrna_mirna_df['qend_miRNA'][i] - 1:end_ind])\n",
    "\n",
    "    else: # miRNA is on the left of mRNA\n",
    "        start_ind = 0\n",
    "        if mrna_mirna_df['gap'][i] <= 0:\n",
    "            qual_flag = check_quality(qual_seq[start_ind:mrna_mirna_df['qstart_miRNA'][i]])\n",
    "        else:\n",
    "            end_ind = mrna_mirna_df['qend_miRNA'][i] + int(mrna_mirna_df['gap'][i])\n",
    "            qual_flag = check_quality(qual_seq[mrna_mirna_df['qend_miRNA'][i] - 1:end_ind])\n",
    "            if qual_flag == 0:\n",
    "                qual_flag = check_quality(qual_seq[start_ind:mrna_mirna_df['qstart_miRNA'][i]])\n",
    "\n",
    "    return qual_flag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mismatch(chim_mirna, ori_mirna, quality_score):\n",
    "    if len(chim_mirna) != len(quality_score):\n",
    "        raise ValueError(\"The length of chim_mirna does not match the length of quality_sequences\")\n",
    "    \n",
    "    if len(chim_mirna) != len(ori_mirna):\n",
    "        raise ValueError(\"The length of chim_mirna does not match the length of ori_mirna\")\n",
    "    \n",
    "    for i in range(len(chim_mirna)):\n",
    "        if chim_mirna[i] != ori_mirna[i]:\n",
    "            if quality_score[i] < 30:\n",
    "                return 1\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df['annotation'] = \"\"\n",
    "mrna_mirna_df['final_seq'] = \"\"\n",
    "mrna_mirna_df['chimeric_seq'] = \"\"\n",
    "\n",
    "for i in range(len(mrna_mirna_df)):\n",
    "    chim_mirna = mrna_mirna_df['qseq_miRNA'][i] # aligned miRNA portion from chimeric sequence\n",
    "    ori_mirna = sequences[mrna_mirna_df['saccver_miRNA'][i]] #original miRNA sequence from miRBASE\n",
    "    ori_mirna_portion = mrna_mirna_df['sseq_miRNA'][i] #aligned miRNA portion from original miRNA sequence\n",
    "    seq_id = mrna_mirna_df['qaccver'][i]\n",
    "    quality_score = chim_qual_seq[seq_id]\n",
    "    chim_seq = ori_seq[seq_id]\n",
    "    front_gap = mrna_mirna_df['sstart_miRNA'][i] - 1\n",
    "    back_gap = mrna_mirna_df['slen_miRNA'][i] - mrna_mirna_df['send_miRNA'][i]\n",
    "    addition_flag = 0\n",
    "    annotation = \"\"\n",
    "\n",
    "    if front_gap > 4 | back_gap > 4:\n",
    "        annotation = 'nothing' \n",
    "        mrna_mirna_df.at[i, 'annotation'] = annotation\n",
    "        mrna_mirna_df.at[i, 'final_seq'] = chim_mirna\n",
    "        continue\n",
    "\n",
    "    if chim_mirna == ori_mirna:\n",
    "        final_seq = mrna_mirna_df['spec_seq'][i]\n",
    "        if len(final_seq) == len(ori_mirna):\n",
    "            annotation = 'miRNA-mRNA'\n",
    "        else:\n",
    "            qual_flag = get_qual_flag(seq_id, i)\n",
    "            if qual_flag == 1:\n",
    "                annotation = 'miRNA-mRNA'\n",
    "            else:\n",
    "                if len(mrna_mirna_df[mrna_mirna_df['spec_seq'] == final_seq]) >= 10:\n",
    "                    annotation = \"isomiR-mRNA\"\n",
    "                    addition_flag = 1\n",
    "                else:\n",
    "                    annotation = \"miRNA-mRNA\"\n",
    "  \n",
    "    else:\n",
    "        if len(chim_mirna) == len(ori_mirna):\n",
    "            if mismatch(chim_mirna, ori_mirna, quality_score) == 0:\n",
    "                if len(mrna_mirna_df[mrna_mirna_df['qseq_miRNA'] == chim_mirna]) >= 10:\n",
    "                    annotation = 'isomiR-mRNA'\n",
    "                else:\n",
    "                    annotation = 'nothing'\n",
    "            else:\n",
    "                annotation = 'nothing'\n",
    "\n",
    "        elif len(chim_mirna) < len(ori_mirna):\n",
    "            if mismatch(chim_mirna, ori_mirna_portion, quality_score) == 0:\n",
    "                if len(mrna_mirna_df[mrna_mirna_df['qseq_miRNA'] == chim_mirna]) >= 10:\n",
    "                    annotation = 'isomiR-mRNA'\n",
    "                else:\n",
    "                    annotation = 'nothing'\n",
    "\n",
    "            else:\n",
    "                annotation = 'nothing'\n",
    "\n",
    "    mrna_mirna_df.at[i, 'annotation'] = annotation  \n",
    "\n",
    "    if addition_flag == 1:\n",
    "        mrna_mirna_df.at[i, 'final_seq'] = final_seq\n",
    "    else:\n",
    "        mrna_mirna_df.at[i, 'final_seq'] = chim_mirna\n",
    "\n",
    "    mrna_mirna_df.at[i, 'chimeric_seq'] = chim_seq  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "annotation\n",
       "isomiR-mRNA    41\n",
       "nothing         8\n",
       "miRNA-mRNA      3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mrna_mirna_df['annotation'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#checking if there is addition in Chimeric Sequences. If there is Additon then need to modify next block of code\n",
    "for i in range(len(mrna_mirna_df)):\n",
    "    chim_mirna = mrna_mirna_df['final_seq'][i] # aligned miRNA portion from chimeric sequence\n",
    "    ori_mirna = sequences[mrna_mirna_df['saccver_miRNA'][i]] #original miRNA sequence from miRBASE\n",
    "    if len(chim_mirna) > len(ori_mirna):\n",
    "        print(\"YES\")\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "mrna_mirna_df.to_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\fields\\SRR14432308\\final_mrna_mirna_annotation.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qaccver</th>\n",
       "      <th>qlen_mRNA</th>\n",
       "      <th>saccver_mRNA</th>\n",
       "      <th>slen_mRNA</th>\n",
       "      <th>qstart_mRNA</th>\n",
       "      <th>qend_mRNA</th>\n",
       "      <th>sstart_mRNA</th>\n",
       "      <th>send_mRNA</th>\n",
       "      <th>qseq_mRNA</th>\n",
       "      <th>sseq_mRNA</th>\n",
       "      <th>...</th>\n",
       "      <th>evalue_miRNA</th>\n",
       "      <th>bitscore_miRNA</th>\n",
       "      <th>length_miRNA</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gap</th>\n",
       "      <th>overlap</th>\n",
       "      <th>spec_seq</th>\n",
       "      <th>annotation</th>\n",
       "      <th>final_seq</th>\n",
       "      <th>chimeric_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR14432308.10833827</td>\n",
       "      <td>58</td>\n",
       "      <td>ENST00000554172.2|ENSG00000075891.24|OTTHUMG00...</td>\n",
       "      <td>1755</td>\n",
       "      <td>22</td>\n",
       "      <td>58</td>\n",
       "      <td>1710</td>\n",
       "      <td>1746</td>\n",
       "      <td>GCAACGTGTCAGGCACACAGACATACCCAGTTGTGAC</td>\n",
       "      <td>GCAACGTGTCAGGCACACAGACATACCCAGTTGTGAC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGCAACGTGTCAGGCACACAGACATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR14432308.11361384</td>\n",
       "      <td>64</td>\n",
       "      <td>ENST00000513145.1|ENSG00000114019.15|OTTHUMG00...</td>\n",
       "      <td>2653</td>\n",
       "      <td>16</td>\n",
       "      <td>64</td>\n",
       "      <td>1448</td>\n",
       "      <td>1496</td>\n",
       "      <td>ACGATCAGCAGCAGGAGCAAGAGAAGCTGGAGCGAGAGATGGCACTGCT</td>\n",
       "      <td>ACGAACAGCAGCAGGAGCAAGAGAAGCTGGAGCGAGAGATGGCACTGCT</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>2</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAGCAGCAGGAGCAAGAGAAGCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR14432308.11391417</td>\n",
       "      <td>54</td>\n",
       "      <td>ENST00000300843.8|ENSG00000007047.16|OTTHUMG00...</td>\n",
       "      <td>5209</td>\n",
       "      <td>21</td>\n",
       "      <td>54</td>\n",
       "      <td>2806</td>\n",
       "      <td>2839</td>\n",
       "      <td>CAGAAGAAGGATGAGGGGGCTCAGCGGGGGGAGC</td>\n",
       "      <td>CAGAAGAAGGATGAGGGGGCTCAGCGGGGGGAGC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAGAAGAAGGATGAGGGGGCTCAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR14432308.11718997</td>\n",
       "      <td>120</td>\n",
       "      <td>ENST00000576541.2|ENSG00000185624.16|OTTHUMG00...</td>\n",
       "      <td>1285</td>\n",
       "      <td>1</td>\n",
       "      <td>99</td>\n",
       "      <td>661</td>\n",
       "      <td>759</td>\n",
       "      <td>AGCTGGCAGAGAGGCTGATGACATCGTGAACTGGCTGAAGAAGCGC...</td>\n",
       "      <td>AGCTGGCAGAGAGGCTGATGACATCGTGAACTGGCTGAAGAAGCGC...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.066</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>CAGAGTTCTACAGTCCGGCCT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>AGCTGGCAGAGAGGCTGATGACATCGTGAACTGGCTGAAGAAGCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR14432308.12758432</td>\n",
       "      <td>53</td>\n",
       "      <td>ENST00000160740.7|ENSG00000079432.9|OTTHUMG000...</td>\n",
       "      <td>6111</td>\n",
       "      <td>15</td>\n",
       "      <td>47</td>\n",
       "      <td>406</td>\n",
       "      <td>438</td>\n",
       "      <td>GACGATCCGCGCGCGGGTCGGGGGGCGGGGCGG</td>\n",
       "      <td>GACGGCCCCCGCGCGCGCCGGGGGGCGGGGCGG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCCGCGCGCGGGTCGGGGGGCGGGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR14432308.13911193</td>\n",
       "      <td>72</td>\n",
       "      <td>ENST00000539782.1|ENSG00000111716.14|OTTHUMG00...</td>\n",
       "      <td>487</td>\n",
       "      <td>22</td>\n",
       "      <td>72</td>\n",
       "      <td>194</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGGAAAAACTCATTGCACCAGTTGCGGAAGAAGAGGCAACAGTTC...</td>\n",
       "      <td>AAGGAAAAACTCATTGCACCAGTTGCGGAAGAAGAGGCAACAGTTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAAGGAAAAACTCATTGCACCAGTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR14432308.14240978</td>\n",
       "      <td>110</td>\n",
       "      <td>ENST00000424544.2|ENSG00000204580.14|OTTHUMG00...</td>\n",
       "      <td>587</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>247</td>\n",
       "      <td>336</td>\n",
       "      <td>GGTGGGCACCCAGGGACGGCATGCCGGGGGCCTGGGCAAGGAGTTC...</td>\n",
       "      <td>GGTGGGCACCCAGGGACGGCATGCCGGGGGCCTGGGCAAGGAGTTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>CAGAGTTCTACAGTCCGCCT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GGTGGGCACCCAGGGACGGCATGCCGGGGGCCTGGGCAAGGAGTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR14432308.14310303</td>\n",
       "      <td>50</td>\n",
       "      <td>ENST00000684552.1|ENSG00000171503.13|OTTHUMG00...</td>\n",
       "      <td>6777</td>\n",
       "      <td>22</td>\n",
       "      <td>50</td>\n",
       "      <td>1746</td>\n",
       "      <td>1774</td>\n",
       "      <td>TGGTGAAACCCCGTCTCTACTAAAAATAC</td>\n",
       "      <td>TGGTGAAACCCCGTCTCTACTAAAAATAC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACCAAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACCAATTGGTGAAACCCCGTCTCTACTAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR14432308.14387577</td>\n",
       "      <td>66</td>\n",
       "      <td>ENST00000643543.1|ENSG00000178971.17|OTTHUMG00...</td>\n",
       "      <td>4859</td>\n",
       "      <td>43</td>\n",
       "      <td>65</td>\n",
       "      <td>4754</td>\n",
       "      <td>4776</td>\n",
       "      <td>TAACGCCAAGGTCGCGGGTTCGA</td>\n",
       "      <td>TAACGCCAAGGTCGCGGGTTCGA</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGACGATCTAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR14432308.14767795</td>\n",
       "      <td>114</td>\n",
       "      <td>ENST00000423214.1|ENSG00000182871.16|OTTHUMG00...</td>\n",
       "      <td>839</td>\n",
       "      <td>75</td>\n",
       "      <td>97</td>\n",
       "      <td>181</td>\n",
       "      <td>203</td>\n",
       "      <td>AGAACGGGTTCCGGAAGGTCCAG</td>\n",
       "      <td>AGAACGGGTTCCGGAAGGTCCAG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.062</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TTCAGAGTTCTACAGTC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>ACCGCGCGGCCCTGGCGGGCAGATGACATCCTGGCCAGCCCCCCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR14432308.2097013</td>\n",
       "      <td>84</td>\n",
       "      <td>ENST00000263620.8|ENSG00000116017.11|OTTHUMG00...</td>\n",
       "      <td>5948</td>\n",
       "      <td>21</td>\n",
       "      <td>84</td>\n",
       "      <td>302</td>\n",
       "      <td>365</td>\n",
       "      <td>CAGGCCGTCATGGAGACGCTGTTGCAGCGGCAGCAGCGGGCGCGCC...</td>\n",
       "      <td>CAGGCCGTGATGGAGACGCTGTTGCAGCGGCAGCAGCGGGCGCGCC...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAGGCCGTCATGGAGACGCTGTTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR14432308.2356280</td>\n",
       "      <td>73</td>\n",
       "      <td>ENST00000507176.5|ENSG00000138814.17|OTTHUMG00...</td>\n",
       "      <td>1958</td>\n",
       "      <td>22</td>\n",
       "      <td>73</td>\n",
       "      <td>1033</td>\n",
       "      <td>1084</td>\n",
       "      <td>AAGGAACAAGATCCGAGCAATAGGCAAAATGGCCAGAGTGTTCTCA...</td>\n",
       "      <td>AAGGAACAAGATCCGAGCAATAGGCAAAATGGCCAGAGTGTTCTCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAAGGAACAAGATCCGAGCAATAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR14432308.2594765</td>\n",
       "      <td>85</td>\n",
       "      <td>ENST00000397536.7|ENSG00000188549.13|OTTHUMG00...</td>\n",
       "      <td>5299</td>\n",
       "      <td>19</td>\n",
       "      <td>62</td>\n",
       "      <td>5100</td>\n",
       "      <td>5143</td>\n",
       "      <td>ATCGTCTACGGCCATACCACCCTGAACGCGCCCGATCTCGTCAG</td>\n",
       "      <td>ATAGTCTACGGCCATACCACCCTGAACGTGCCTAATCTCGTAAG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACG</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGTCTACGGCCATACCACCCTGAACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR14432308.3780940</td>\n",
       "      <td>50</td>\n",
       "      <td>ENST00000706905.1|ENSG00000206503.15|OTTHUMG00...</td>\n",
       "      <td>2361</td>\n",
       "      <td>20</td>\n",
       "      <td>50</td>\n",
       "      <td>2089</td>\n",
       "      <td>2119</td>\n",
       "      <td>GATCCAGATTGTGACATTTGCAGAGAGGAGG</td>\n",
       "      <td>GAACCAGATTGTGACATTTGCAGAGAGGAGG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CAGAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>CAGAGTTCTACAGTCCGACGATCCAGATTGTGACATTTGCAGAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR14432308.4395690</td>\n",
       "      <td>87</td>\n",
       "      <td>ENST00000311506.10|ENSG00000087995.16|OTTHUMG0...</td>\n",
       "      <td>5799</td>\n",
       "      <td>18</td>\n",
       "      <td>62</td>\n",
       "      <td>1035</td>\n",
       "      <td>1079</td>\n",
       "      <td>GATCCTGGTGGATCGCCGACTGCAGGTGAACCGAGGAAAGCAACT</td>\n",
       "      <td>GAACCTGGTGGATCGCCGACTGCAGGTGAACCGAGGAAAGCAACT</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCCTGGTGGATCGCCGACTGCAGGTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR14432308.4470118</td>\n",
       "      <td>120</td>\n",
       "      <td>ENST00000457607.1|ENSG00000132153.15|OTTHUMG00...</td>\n",
       "      <td>3748</td>\n",
       "      <td>1</td>\n",
       "      <td>101</td>\n",
       "      <td>1471</td>\n",
       "      <td>1571</td>\n",
       "      <td>ACGGGCTGTGGGAAGACCACGCGCATCCCCCAGCTGTTGCTGGAGC...</td>\n",
       "      <td>ACGGGCTGTGGGAAGACCACGCGCATCCCCCAGCTGTTGCTGGAGC...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.066</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TTCAGAGTTCTACAGTCCC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>ACGGGCTGTGGGAAGACCACGCGCATCCCCCAGCTGTTGCTGGAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRR14432308.4494772</td>\n",
       "      <td>61</td>\n",
       "      <td>ENST00000577787.1|ENSG00000108654.16|OTTHUMG00...</td>\n",
       "      <td>838</td>\n",
       "      <td>21</td>\n",
       "      <td>61</td>\n",
       "      <td>148</td>\n",
       "      <td>188</td>\n",
       "      <td>CGTGATTTGGAGAGAGGTGTGGAAATCTGTATTGCAACACC</td>\n",
       "      <td>CGTGATTTGGAGAGAGGTGTGGAAATCTGTATTGCAACACC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGTGATTTGGAGAGAGGTGTGGAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR14432308.4854069</td>\n",
       "      <td>74</td>\n",
       "      <td>ENST00000605930.3|ENSG00000096433.11|OTTHUMG00...</td>\n",
       "      <td>9079</td>\n",
       "      <td>23</td>\n",
       "      <td>74</td>\n",
       "      <td>5492</td>\n",
       "      <td>5543</td>\n",
       "      <td>ATGCGACCTCATCACCAGCACCAAGAACGAGAAGATCTTCCAGGAG...</td>\n",
       "      <td>ATGCGACCTCATCACCAGCACCAAGAACGAGAAGATCTTCCAGGAG...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>AGAGTTCTACAGTCCGACGATCATGCGACCTCATCACCAGCACCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR14432308.4958848</td>\n",
       "      <td>55</td>\n",
       "      <td>ENST00000575870.6|ENSG00000167978.19|OTTHUMG00...</td>\n",
       "      <td>4432</td>\n",
       "      <td>23</td>\n",
       "      <td>55</td>\n",
       "      <td>1625</td>\n",
       "      <td>1657</td>\n",
       "      <td>AGGGGTAGATCTCGTTCTAGAACACCAGCCCGC</td>\n",
       "      <td>AGGGGTAGATCTCGTTCTAGAACACCAGCCCGC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>AGAGTTCTACAGTCCGACGATCAGGGGTAGATCTCGTTCTAGAACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR14432308.518630</td>\n",
       "      <td>51</td>\n",
       "      <td>ENST00000512309.2|ENSG00000184432.11|OTTHUMG00...</td>\n",
       "      <td>3373</td>\n",
       "      <td>2</td>\n",
       "      <td>34</td>\n",
       "      <td>3201</td>\n",
       "      <td>3233</td>\n",
       "      <td>TTGAATGCTACCCTGACCACAGTGCTTTGGACT</td>\n",
       "      <td>TTGATTGCTACCCTGACCACAGTGCTTTGGACT</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TTCCGAGTTCTACAGTC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>TTTGAATGCTACCCTGACCACAGTGCTTTGGACTTTCCGAGTTCTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRR14432308.525236</td>\n",
       "      <td>65</td>\n",
       "      <td>ENST00000528760.1|ENSG00000149257.16|OTTHUMG00...</td>\n",
       "      <td>773</td>\n",
       "      <td>22</td>\n",
       "      <td>65</td>\n",
       "      <td>567</td>\n",
       "      <td>610</td>\n",
       "      <td>ACGGACCCAGCTCAGTGAGCTTCGCTGATGACTTCGTGCGCAGC</td>\n",
       "      <td>ACGGACCCAGCTCAGTGAGCTTCGCTGATGACTTCGTGCGCAGC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCACGGACCCAGCTCAGTGAGCTTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR14432308.5609886</td>\n",
       "      <td>78</td>\n",
       "      <td>ENST00000562337.5|ENSG00000103495.15|OTTHUMG00...</td>\n",
       "      <td>1574</td>\n",
       "      <td>23</td>\n",
       "      <td>42</td>\n",
       "      <td>318</td>\n",
       "      <td>337</td>\n",
       "      <td>CAGGGCTGCGCCCAGAGTCC</td>\n",
       "      <td>CAGGGCTGCGCCCAGAGTCC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>CAGAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>CAGAGTTCTACAGTCCGACGATCAGGGCTGCGCCCAGAGTCCCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR14432308.5843381</td>\n",
       "      <td>58</td>\n",
       "      <td>ENST00000525844.1|ENSG00000145734.20|OTTHUMG00...</td>\n",
       "      <td>4997</td>\n",
       "      <td>22</td>\n",
       "      <td>45</td>\n",
       "      <td>1742</td>\n",
       "      <td>1765</td>\n",
       "      <td>CTGATCTGCTTCCATCTCCAACTG</td>\n",
       "      <td>CTGATCTGCTTCCATCTCCAAGTG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCCTGATCTGCTTCCATCTCCAACTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR14432308.6050043</td>\n",
       "      <td>119</td>\n",
       "      <td>ENST00000534779.5|ENSG00000089597.18|OTTHUMG00...</td>\n",
       "      <td>2626</td>\n",
       "      <td>1</td>\n",
       "      <td>101</td>\n",
       "      <td>1663</td>\n",
       "      <td>1763</td>\n",
       "      <td>TTGGGGCTGGTGGGACTTTCCTTCTGTGGGGCGGATGTGGGTGGCT...</td>\n",
       "      <td>TTGGGGCTGGTGGGACTTTCCTTCTGTGGGGCGGATGTGGGTGGCT...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.066</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TTCAGAGTTCTACAGTCC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>TTGGGGCTGGTGGGACTTTCCTTCTGTGGGGCGGATGTGGGTGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SRR14432308.6198142</td>\n",
       "      <td>122</td>\n",
       "      <td>ENST00000377269.3|ENSG00000130985.17|OTTHUMG00...</td>\n",
       "      <td>2497</td>\n",
       "      <td>79</td>\n",
       "      <td>102</td>\n",
       "      <td>1150</td>\n",
       "      <td>1173</td>\n",
       "      <td>CACTGCGGCCGGCAGGCACTCAGC</td>\n",
       "      <td>CACTGCGGCTGGCAGGCACTCAGC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>TCAGAGTTCTACAGTCCCCT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>GTGCAGCGCAGCCTGGTGCTGCAGCGACCACAGACCTGGGCTGACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SRR14432308.6496907</td>\n",
       "      <td>62</td>\n",
       "      <td>ENST00000261609.13|ENSG00000128731.18|OTTHUMG0...</td>\n",
       "      <td>15364</td>\n",
       "      <td>21</td>\n",
       "      <td>62</td>\n",
       "      <td>7283</td>\n",
       "      <td>7324</td>\n",
       "      <td>CAGCTGCTGGCCCCGGCCACCCAGCCGTCTCCTGTGAAGGCC</td>\n",
       "      <td>CAGCTGCTGGCCTCGGCCACCCAGCCGTCTCCTGTGAAGGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAGCTGCTGGCCCCGGCCACCCAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SRR14432308.6825834</td>\n",
       "      <td>69</td>\n",
       "      <td>ENST00000361791.7|ENSG00000075891.24|OTTHUMG00...</td>\n",
       "      <td>3193</td>\n",
       "      <td>22</td>\n",
       "      <td>69</td>\n",
       "      <td>919</td>\n",
       "      <td>966</td>\n",
       "      <td>GCAACGTGTCAGGCACACAGACATACCCAGTTGTGACTGGTCGTGACA</td>\n",
       "      <td>GCAACGTGTCAGGCACACAGACATACCCAGTTGTGACTGGTCGTGACA</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGCAACGTGTCAGGCACACAGACATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SRR14432308.693916</td>\n",
       "      <td>76</td>\n",
       "      <td>ENST00000562337.5|ENSG00000103495.15|OTTHUMG00...</td>\n",
       "      <td>1574</td>\n",
       "      <td>21</td>\n",
       "      <td>40</td>\n",
       "      <td>318</td>\n",
       "      <td>337</td>\n",
       "      <td>CAGGGCTGCGCCCAGAGTCC</td>\n",
       "      <td>CAGGGCTGCGCCCAGAGTCC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAGGGCTGCGCCCAGAGTCCCAGGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SRR14432308.7108693</td>\n",
       "      <td>51</td>\n",
       "      <td>ENST00000525576.1|ENSG00000162231.14|OTTHUMG00...</td>\n",
       "      <td>572</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>109</td>\n",
       "      <td>132</td>\n",
       "      <td>AGCGAACACGATGATGAACGCGTT</td>\n",
       "      <td>AGCGAACACGATGATGAACGCGTT</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>CAGAGTTCTACAGTCCGACGAATGCCT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>AGCGAACACGATGATGAACGCGTTCAGAGTTCTACAGTCCGACGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SRR14432308.7577210</td>\n",
       "      <td>59</td>\n",
       "      <td>ENST00000366956.10|ENSG00000152104.12|OTTHUMG0...</td>\n",
       "      <td>13360</td>\n",
       "      <td>22</td>\n",
       "      <td>59</td>\n",
       "      <td>1209</td>\n",
       "      <td>1246</td>\n",
       "      <td>TGGATGGATTTGGACAGGAAATCTTCCCTGTAAAGGAC</td>\n",
       "      <td>TGGATGGATTTGGACAGGAAATCTTCCCTGTAAAGGAC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCTGGATGGATTTGGACAGGAAATCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SRR14432308.7693614</td>\n",
       "      <td>118</td>\n",
       "      <td>ENST00000521569.1|ENSG00000113140.11|OTTHUMG00...</td>\n",
       "      <td>602</td>\n",
       "      <td>1</td>\n",
       "      <td>101</td>\n",
       "      <td>402</td>\n",
       "      <td>502</td>\n",
       "      <td>AAGCTCCACCTGGACTACATCGGGCCTTGCAAATACATCCCCCCTT...</td>\n",
       "      <td>AAGCTCCACCTGGACTACATCGGGCCTTGCAAATACATCCCCCCTT...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.065</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TTCAGAGTTCTACAGTC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>AAGCTCCACCTGGACTACATCGGGCCTTGCAAATACATCCCCCCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SRR14432308.7817471</td>\n",
       "      <td>83</td>\n",
       "      <td>ENST00000366956.10|ENSG00000152104.12|OTTHUMG0...</td>\n",
       "      <td>13360</td>\n",
       "      <td>4</td>\n",
       "      <td>67</td>\n",
       "      <td>36</td>\n",
       "      <td>99</td>\n",
       "      <td>GGCGCAGACTCGGGGGCAGGTTTTTGTGTAACGCTGCGGTCACCAC...</td>\n",
       "      <td>GGCGCAGACTCGGGGGCAGGTTGCTGTGCTTCTCCGGGCTCAGCCG...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGTTCTACAGTCCCCT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>AGCGGCGCAGACTCGGGGGCAGGTTTTTGTGTAACGCTGCGGTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SRR14432308.8229069</td>\n",
       "      <td>67</td>\n",
       "      <td>ENST00000402641.6|ENSG00000059145.19|OTTHUMG00...</td>\n",
       "      <td>669</td>\n",
       "      <td>21</td>\n",
       "      <td>67</td>\n",
       "      <td>517</td>\n",
       "      <td>563</td>\n",
       "      <td>CAAACGGAGCTGAGCTGGCCTGGGTCAGGCGGCAGCTGGACGAGGCC</td>\n",
       "      <td>CAAACGGAGCTGAGCTGGCCCGGGTCAGGCGGCAGCTGGACGAGGCC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAAACGGAGCTGAGCTGGCCTGGGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SRR14432308.8331317</td>\n",
       "      <td>133</td>\n",
       "      <td>ENST00000619601.1|ENSG00000111640.15|OTTHUMG00...</td>\n",
       "      <td>1086</td>\n",
       "      <td>1</td>\n",
       "      <td>54</td>\n",
       "      <td>130</td>\n",
       "      <td>183</td>\n",
       "      <td>AAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACCGAGA...</td>\n",
       "      <td>AAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACTGGCG...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074</td>\n",
       "      <td>24.7</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TTCAGAGTTCTACAGTCCTTTTTTTTTTTAATGATACGGCGACCAC...</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTC</td>\n",
       "      <td>AAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACCGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SRR14432308.8416224</td>\n",
       "      <td>53</td>\n",
       "      <td>ENST00000396037.9|ENSG00000111728.11|OTTHUMG00...</td>\n",
       "      <td>9707</td>\n",
       "      <td>21</td>\n",
       "      <td>53</td>\n",
       "      <td>7534</td>\n",
       "      <td>7566</td>\n",
       "      <td>CACTCCGGATGTGCTGACCCCTGCGATTTCCCC</td>\n",
       "      <td>CACACCTGATGTGCTGGCCCCTGCGATTTCCCC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCACTCCGGATGTGCTGACCCCTGCGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>SRR14432308.8463167</td>\n",
       "      <td>52</td>\n",
       "      <td>ENST00000160740.7|ENSG00000079432.9|OTTHUMG000...</td>\n",
       "      <td>6111</td>\n",
       "      <td>15</td>\n",
       "      <td>47</td>\n",
       "      <td>406</td>\n",
       "      <td>438</td>\n",
       "      <td>GACGATCCGCGCGCGGGTCGGGGGGCGGGGCGG</td>\n",
       "      <td>GACGGCCCCCGCGCGCGCCGGGGGGCGGGGCGG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCCGCGCGCGGGTCGGGGGGCGGGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SRR14432308.8981173</td>\n",
       "      <td>63</td>\n",
       "      <td>ENST00000263269.4|ENSG00000105464.4|OTTHUMG000...</td>\n",
       "      <td>5511</td>\n",
       "      <td>21</td>\n",
       "      <td>48</td>\n",
       "      <td>3466</td>\n",
       "      <td>3493</td>\n",
       "      <td>CAGCCGGTCCCCCGCTGTCCCCGCCGGC</td>\n",
       "      <td>CAGCCGGGCGCCCGCTGTCCCCGCCGGC</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCAGCCGGTCCCCCGCTGTCCCCGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SRR14432308.9149176</td>\n",
       "      <td>85</td>\n",
       "      <td>ENST00000602198.5|ENSG00000131126.19|OTTHUMG00...</td>\n",
       "      <td>1446</td>\n",
       "      <td>21</td>\n",
       "      <td>83</td>\n",
       "      <td>267</td>\n",
       "      <td>329</td>\n",
       "      <td>CGAATGGATGAACGAGATTCCCACTGTCCCTACCTACTATCCAGCG...</td>\n",
       "      <td>CGAATGGATAGACGCTATTCCCACCGTCCCTACATAGCATCCAGCG...</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGAATGGATGAACGAGATTCCCACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SRR14432308.9663750</td>\n",
       "      <td>74</td>\n",
       "      <td>ENST00000160740.7|ENSG00000079432.9|OTTHUMG000...</td>\n",
       "      <td>6111</td>\n",
       "      <td>36</td>\n",
       "      <td>68</td>\n",
       "      <td>406</td>\n",
       "      <td>438</td>\n",
       "      <td>GACGATCCGCGCGCGGGTCGGGGGGCGGGGCGG</td>\n",
       "      <td>GACGGCCCCCGCGCGCGCCGGGGGGCGGGGCGG</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>3</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGACGATCCGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>SRR14432308.9819424</td>\n",
       "      <td>67</td>\n",
       "      <td>ENST00000620529.1|ENSG00000185619.19|OTTHUMG00...</td>\n",
       "      <td>1899</td>\n",
       "      <td>18</td>\n",
       "      <td>43</td>\n",
       "      <td>1685</td>\n",
       "      <td>1710</td>\n",
       "      <td>GATCACGAAGATACATGTGTTTGACT</td>\n",
       "      <td>GATCACGAAGATACATGTGTTTGACT</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCACGAAGATACATGTGTTTGACTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SRR14432308.9926460</td>\n",
       "      <td>62</td>\n",
       "      <td>ENST00000422840.7|ENSG00000090863.12|OTTHUMG00...</td>\n",
       "      <td>9287</td>\n",
       "      <td>21</td>\n",
       "      <td>62</td>\n",
       "      <td>4118</td>\n",
       "      <td>4159</td>\n",
       "      <td>CCGAGGCGTGGGTGGGCAGCCGTTCCCCAGGCTGTGTGGGGT</td>\n",
       "      <td>CCGAGGCGTGGGTGGGCAGCCGTTCCCCAGGCTGTGTGGGGT</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010</td>\n",
       "      <td>27.4</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>GAGTTCTACAGTCCGACGAT</td>\n",
       "      <td>isomiR-mRNA</td>\n",
       "      <td>GAGTTCTACAGTCCGAC</td>\n",
       "      <td>GAGTTCTACAGTCCGACGATCCGAGGCGTGGGTGGGCAGCCGTTCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 qaccver  qlen_mRNA  \\\n",
       "0   SRR14432308.10833827         58   \n",
       "1   SRR14432308.11361384         64   \n",
       "2   SRR14432308.11391417         54   \n",
       "3   SRR14432308.11718997        120   \n",
       "4   SRR14432308.12758432         53   \n",
       "5   SRR14432308.13911193         72   \n",
       "6   SRR14432308.14240978        110   \n",
       "7   SRR14432308.14310303         50   \n",
       "8   SRR14432308.14387577         66   \n",
       "9   SRR14432308.14767795        114   \n",
       "10   SRR14432308.2097013         84   \n",
       "11   SRR14432308.2356280         73   \n",
       "12   SRR14432308.2594765         85   \n",
       "13   SRR14432308.3780940         50   \n",
       "14   SRR14432308.4395690         87   \n",
       "15   SRR14432308.4470118        120   \n",
       "16   SRR14432308.4494772         61   \n",
       "17   SRR14432308.4854069         74   \n",
       "18   SRR14432308.4958848         55   \n",
       "19    SRR14432308.518630         51   \n",
       "20    SRR14432308.525236         65   \n",
       "21   SRR14432308.5609886         78   \n",
       "22   SRR14432308.5843381         58   \n",
       "23   SRR14432308.6050043        119   \n",
       "24   SRR14432308.6198142        122   \n",
       "25   SRR14432308.6496907         62   \n",
       "26   SRR14432308.6825834         69   \n",
       "27    SRR14432308.693916         76   \n",
       "28   SRR14432308.7108693         51   \n",
       "29   SRR14432308.7577210         59   \n",
       "30   SRR14432308.7693614        118   \n",
       "31   SRR14432308.7817471         83   \n",
       "32   SRR14432308.8229069         67   \n",
       "33   SRR14432308.8331317        133   \n",
       "34   SRR14432308.8416224         53   \n",
       "35   SRR14432308.8463167         52   \n",
       "36   SRR14432308.8981173         63   \n",
       "37   SRR14432308.9149176         85   \n",
       "38   SRR14432308.9663750         74   \n",
       "39   SRR14432308.9819424         67   \n",
       "40   SRR14432308.9926460         62   \n",
       "\n",
       "                                         saccver_mRNA  slen_mRNA  qstart_mRNA  \\\n",
       "0   ENST00000554172.2|ENSG00000075891.24|OTTHUMG00...       1755           22   \n",
       "1   ENST00000513145.1|ENSG00000114019.15|OTTHUMG00...       2653           16   \n",
       "2   ENST00000300843.8|ENSG00000007047.16|OTTHUMG00...       5209           21   \n",
       "3   ENST00000576541.2|ENSG00000185624.16|OTTHUMG00...       1285            1   \n",
       "4   ENST00000160740.7|ENSG00000079432.9|OTTHUMG000...       6111           15   \n",
       "5   ENST00000539782.1|ENSG00000111716.14|OTTHUMG00...        487           22   \n",
       "6   ENST00000424544.2|ENSG00000204580.14|OTTHUMG00...        587            1   \n",
       "7   ENST00000684552.1|ENSG00000171503.13|OTTHUMG00...       6777           22   \n",
       "8   ENST00000643543.1|ENSG00000178971.17|OTTHUMG00...       4859           43   \n",
       "9   ENST00000423214.1|ENSG00000182871.16|OTTHUMG00...        839           75   \n",
       "10  ENST00000263620.8|ENSG00000116017.11|OTTHUMG00...       5948           21   \n",
       "11  ENST00000507176.5|ENSG00000138814.17|OTTHUMG00...       1958           22   \n",
       "12  ENST00000397536.7|ENSG00000188549.13|OTTHUMG00...       5299           19   \n",
       "13  ENST00000706905.1|ENSG00000206503.15|OTTHUMG00...       2361           20   \n",
       "14  ENST00000311506.10|ENSG00000087995.16|OTTHUMG0...       5799           18   \n",
       "15  ENST00000457607.1|ENSG00000132153.15|OTTHUMG00...       3748            1   \n",
       "16  ENST00000577787.1|ENSG00000108654.16|OTTHUMG00...        838           21   \n",
       "17  ENST00000605930.3|ENSG00000096433.11|OTTHUMG00...       9079           23   \n",
       "18  ENST00000575870.6|ENSG00000167978.19|OTTHUMG00...       4432           23   \n",
       "19  ENST00000512309.2|ENSG00000184432.11|OTTHUMG00...       3373            2   \n",
       "20  ENST00000528760.1|ENSG00000149257.16|OTTHUMG00...        773           22   \n",
       "21  ENST00000562337.5|ENSG00000103495.15|OTTHUMG00...       1574           23   \n",
       "22  ENST00000525844.1|ENSG00000145734.20|OTTHUMG00...       4997           22   \n",
       "23  ENST00000534779.5|ENSG00000089597.18|OTTHUMG00...       2626            1   \n",
       "24  ENST00000377269.3|ENSG00000130985.17|OTTHUMG00...       2497           79   \n",
       "25  ENST00000261609.13|ENSG00000128731.18|OTTHUMG0...      15364           21   \n",
       "26  ENST00000361791.7|ENSG00000075891.24|OTTHUMG00...       3193           22   \n",
       "27  ENST00000562337.5|ENSG00000103495.15|OTTHUMG00...       1574           21   \n",
       "28  ENST00000525576.1|ENSG00000162231.14|OTTHUMG00...        572            1   \n",
       "29  ENST00000366956.10|ENSG00000152104.12|OTTHUMG0...      13360           22   \n",
       "30  ENST00000521569.1|ENSG00000113140.11|OTTHUMG00...        602            1   \n",
       "31  ENST00000366956.10|ENSG00000152104.12|OTTHUMG0...      13360            4   \n",
       "32  ENST00000402641.6|ENSG00000059145.19|OTTHUMG00...        669           21   \n",
       "33  ENST00000619601.1|ENSG00000111640.15|OTTHUMG00...       1086            1   \n",
       "34  ENST00000396037.9|ENSG00000111728.11|OTTHUMG00...       9707           21   \n",
       "35  ENST00000160740.7|ENSG00000079432.9|OTTHUMG000...       6111           15   \n",
       "36  ENST00000263269.4|ENSG00000105464.4|OTTHUMG000...       5511           21   \n",
       "37  ENST00000602198.5|ENSG00000131126.19|OTTHUMG00...       1446           21   \n",
       "38  ENST00000160740.7|ENSG00000079432.9|OTTHUMG000...       6111           36   \n",
       "39  ENST00000620529.1|ENSG00000185619.19|OTTHUMG00...       1899           18   \n",
       "40  ENST00000422840.7|ENSG00000090863.12|OTTHUMG00...       9287           21   \n",
       "\n",
       "    qend_mRNA  sstart_mRNA  send_mRNA  \\\n",
       "0          58         1710       1746   \n",
       "1          64         1448       1496   \n",
       "2          54         2806       2839   \n",
       "3          99          661        759   \n",
       "4          47          406        438   \n",
       "5          72          194        244   \n",
       "6          90          247        336   \n",
       "7          50         1746       1774   \n",
       "8          65         4754       4776   \n",
       "9          97          181        203   \n",
       "10         84          302        365   \n",
       "11         73         1033       1084   \n",
       "12         62         5100       5143   \n",
       "13         50         2089       2119   \n",
       "14         62         1035       1079   \n",
       "15        101         1471       1571   \n",
       "16         61          148        188   \n",
       "17         74         5492       5543   \n",
       "18         55         1625       1657   \n",
       "19         34         3201       3233   \n",
       "20         65          567        610   \n",
       "21         42          318        337   \n",
       "22         45         1742       1765   \n",
       "23        101         1663       1763   \n",
       "24        102         1150       1173   \n",
       "25         62         7283       7324   \n",
       "26         69          919        966   \n",
       "27         40          318        337   \n",
       "28         24          109        132   \n",
       "29         59         1209       1246   \n",
       "30        101          402        502   \n",
       "31         67           36         99   \n",
       "32         67          517        563   \n",
       "33         54          130        183   \n",
       "34         53         7534       7566   \n",
       "35         47          406        438   \n",
       "36         48         3466       3493   \n",
       "37         83          267        329   \n",
       "38         68          406        438   \n",
       "39         43         1685       1710   \n",
       "40         62         4118       4159   \n",
       "\n",
       "                                            qseq_mRNA  \\\n",
       "0               GCAACGTGTCAGGCACACAGACATACCCAGTTGTGAC   \n",
       "1   ACGATCAGCAGCAGGAGCAAGAGAAGCTGGAGCGAGAGATGGCACTGCT   \n",
       "2                  CAGAAGAAGGATGAGGGGGCTCAGCGGGGGGAGC   \n",
       "3   AGCTGGCAGAGAGGCTGATGACATCGTGAACTGGCTGAAGAAGCGC...   \n",
       "4                   GACGATCCGCGCGCGGGTCGGGGGGCGGGGCGG   \n",
       "5   AAGGAAAAACTCATTGCACCAGTTGCGGAAGAAGAGGCAACAGTTC...   \n",
       "6   GGTGGGCACCCAGGGACGGCATGCCGGGGGCCTGGGCAAGGAGTTC...   \n",
       "7                       TGGTGAAACCCCGTCTCTACTAAAAATAC   \n",
       "8                             TAACGCCAAGGTCGCGGGTTCGA   \n",
       "9                             AGAACGGGTTCCGGAAGGTCCAG   \n",
       "10  CAGGCCGTCATGGAGACGCTGTTGCAGCGGCAGCAGCGGGCGCGCC...   \n",
       "11  AAGGAACAAGATCCGAGCAATAGGCAAAATGGCCAGAGTGTTCTCA...   \n",
       "12       ATCGTCTACGGCCATACCACCCTGAACGCGCCCGATCTCGTCAG   \n",
       "13                    GATCCAGATTGTGACATTTGCAGAGAGGAGG   \n",
       "14      GATCCTGGTGGATCGCCGACTGCAGGTGAACCGAGGAAAGCAACT   \n",
       "15  ACGGGCTGTGGGAAGACCACGCGCATCCCCCAGCTGTTGCTGGAGC...   \n",
       "16          CGTGATTTGGAGAGAGGTGTGGAAATCTGTATTGCAACACC   \n",
       "17  ATGCGACCTCATCACCAGCACCAAGAACGAGAAGATCTTCCAGGAG...   \n",
       "18                  AGGGGTAGATCTCGTTCTAGAACACCAGCCCGC   \n",
       "19                  TTGAATGCTACCCTGACCACAGTGCTTTGGACT   \n",
       "20       ACGGACCCAGCTCAGTGAGCTTCGCTGATGACTTCGTGCGCAGC   \n",
       "21                               CAGGGCTGCGCCCAGAGTCC   \n",
       "22                           CTGATCTGCTTCCATCTCCAACTG   \n",
       "23  TTGGGGCTGGTGGGACTTTCCTTCTGTGGGGCGGATGTGGGTGGCT...   \n",
       "24                           CACTGCGGCCGGCAGGCACTCAGC   \n",
       "25         CAGCTGCTGGCCCCGGCCACCCAGCCGTCTCCTGTGAAGGCC   \n",
       "26   GCAACGTGTCAGGCACACAGACATACCCAGTTGTGACTGGTCGTGACA   \n",
       "27                               CAGGGCTGCGCCCAGAGTCC   \n",
       "28                           AGCGAACACGATGATGAACGCGTT   \n",
       "29             TGGATGGATTTGGACAGGAAATCTTCCCTGTAAAGGAC   \n",
       "30  AAGCTCCACCTGGACTACATCGGGCCTTGCAAATACATCCCCCCTT...   \n",
       "31  GGCGCAGACTCGGGGGCAGGTTTTTGTGTAACGCTGCGGTCACCAC...   \n",
       "32    CAAACGGAGCTGAGCTGGCCTGGGTCAGGCGGCAGCTGGACGAGGCC   \n",
       "33  AAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACCGAGA...   \n",
       "34                  CACTCCGGATGTGCTGACCCCTGCGATTTCCCC   \n",
       "35                  GACGATCCGCGCGCGGGTCGGGGGGCGGGGCGG   \n",
       "36                       CAGCCGGTCCCCCGCTGTCCCCGCCGGC   \n",
       "37  CGAATGGATGAACGAGATTCCCACTGTCCCTACCTACTATCCAGCG...   \n",
       "38                  GACGATCCGCGCGCGGGTCGGGGGGCGGGGCGG   \n",
       "39                         GATCACGAAGATACATGTGTTTGACT   \n",
       "40         CCGAGGCGTGGGTGGGCAGCCGTTCCCCAGGCTGTGTGGGGT   \n",
       "\n",
       "                                            sseq_mRNA  ...  evalue_miRNA  \\\n",
       "0               GCAACGTGTCAGGCACACAGACATACCCAGTTGTGAC  ...         0.009   \n",
       "1   ACGAACAGCAGCAGGAGCAAGAGAAGCTGGAGCGAGAGATGGCACTGCT  ...         0.010   \n",
       "2                  CAGAAGAAGGATGAGGGGGCTCAGCGGGGGGAGC  ...         0.008   \n",
       "3   AGCTGGCAGAGAGGCTGATGACATCGTGAACTGGCTGAAGAAGCGC...  ...         0.066   \n",
       "4                   GACGGCCCCCGCGCGCGCCGGGGGGCGGGGCGG  ...         0.008   \n",
       "5   AAGGAAAAACTCATTGCACCAGTTGCGGAAGAAGAGGCAACAGTTC...  ...         0.012   \n",
       "6   GGTGGGCACCCAGGGACGGCATGCCGGGGGCCTGGGCAAGGAGTTC...  ...         0.060   \n",
       "7                       TGGTGAAACCCCGTCTCTACTAAAAATAC  ...         0.007   \n",
       "8                             TAACGCCAAGGTCGCGGGTTCGA  ...         0.011   \n",
       "9                             AGAACGGGTTCCGGAAGGTCCAG  ...         0.062   \n",
       "10  CAGGCCGTGATGGAGACGCTGTTGCAGCGGCAGCAGCGGGCGCGCC...  ...         0.015   \n",
       "11  AAGGAACAAGATCCGAGCAATAGGCAAAATGGCCAGAGTGTTCTCA...  ...         0.012   \n",
       "12       ATAGTCTACGGCCATACCACCCTGAACGTGCCTAATCTCGTAAG  ...         0.015   \n",
       "13                    GAACCAGATTGTGACATTTGCAGAGAGGAGG  ...         0.007   \n",
       "14      GAACCTGGTGGATCGCCGACTGCAGGTGAACCGAGGAAAGCAACT  ...         0.015   \n",
       "15  ACGGGCTGTGGGAAGACCACGCGCATCCCCCAGCTGTTGCTGGAGC...  ...         0.066   \n",
       "16          CGTGATTTGGAGAGAGGTGTGGAAATCTGTATTGCAACACC  ...         0.010   \n",
       "17  ATGCGACCTCATCACCAGCACCAAGAACGAGAAGATCTTCCAGGAG...  ...         0.012   \n",
       "18                  AGGGGTAGATCTCGTTCTAGAACACCAGCCCGC  ...         0.009   \n",
       "19                  TTGATTGCTACCCTGACCACAGTGCTTTGGACT  ...         0.027   \n",
       "20       ACGGACCCAGCTCAGTGAGCTTCGCTGATGACTTCGTGCGCAGC  ...         0.011   \n",
       "21                               CAGGGCTGCGCCCAGAGTCC  ...         0.013   \n",
       "22                           CTGATCTGCTTCCATCTCCAAGTG  ...         0.009   \n",
       "23  TTGGGGCTGGTGGGACTTTCCTTCTGTGGGGCGGATGTGGGTGGCT...  ...         0.066   \n",
       "24                           CACTGCGGCTGGCAGGCACTCAGC  ...         0.067   \n",
       "25         CAGCTGCTGGCCTCGGCCACCCAGCCGTCTCCTGTGAAGGCC  ...         0.010   \n",
       "26   GCAACGTGTCAGGCACACAGACATACCCAGTTGTGACTGGTCGTGACA  ...         0.011   \n",
       "27                               CAGGGCTGCGCCCAGAGTCC  ...         0.013   \n",
       "28                           AGCGAACACGATGATGAACGCGTT  ...         0.008   \n",
       "29             TGGATGGATTTGGACAGGAAATCTTCCCTGTAAAGGAC  ...         0.009   \n",
       "30  AAGCTCCACCTGGACTACATCGGGCCTTGCAAATACATCCCCCCTT...  ...         0.065   \n",
       "31  GGCGCAGACTCGGGGGCAGGTTGCTGTGCTTCTCCGGGCTCAGCCG...  ...         0.050   \n",
       "32    CAAACGGAGCTGAGCTGGCCCGGGTCAGGCGGCAGCTGGACGAGGCC  ...         0.011   \n",
       "33  AAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACTGGCG...  ...         0.074   \n",
       "34                  CACACCTGATGTGCTGGCCCCTGCGATTTCCCC  ...         0.008   \n",
       "35                  GACGGCCCCCGCGCGCGCCGGGGGGCGGGGCGG  ...         0.008   \n",
       "36                       CAGCCGGGCGCCCGCTGTCCCCGCCGGC  ...         0.010   \n",
       "37  CGAATGGATAGACGCTATTCCCACCGTCCCTACATAGCATCCAGCG...  ...         0.015   \n",
       "38                  GACGGCCCCCGCGCGCGCCGGGGGGCGGGGCGG  ...         0.012   \n",
       "39                         GATCACGAAGATACATGTGTTTGACT  ...         0.011   \n",
       "40         CCGAGGCGTGGGTGGGCAGCCGTTCCCCAGGCTGTGTGGGGT  ...         0.010   \n",
       "\n",
       "    bitscore_miRNA  length_miRNA  mismatch gap  overlap  \\\n",
       "0             27.4            17         1   4        0   \n",
       "1             27.4            17         1  -1        2   \n",
       "2             27.4            17         1   3        0   \n",
       "3             24.7            13         0   2        0   \n",
       "4             27.4            17         1  -1        3   \n",
       "5             27.4            17         1   4        0   \n",
       "6             24.7            13         0   2        0   \n",
       "7             27.4            17         1   4        0   \n",
       "8             27.4            17         1   4        0   \n",
       "9             24.7            13         0   4        0   \n",
       "10            27.4            17         1   3        0   \n",
       "11            27.4            17         1   4        0   \n",
       "12            27.4            17         1   1        0   \n",
       "13            27.4            17         1   0        0   \n",
       "14            27.4            17         1   0        0   \n",
       "15            24.7            13         0   4        0   \n",
       "16            27.4            17         1   3        0   \n",
       "17            27.4            17         1   4        0   \n",
       "18            27.4            17         1   4        0   \n",
       "19            24.7            13         0   4        0   \n",
       "20            27.4            17         1   4        0   \n",
       "21            27.4            17         1   3        0   \n",
       "22            27.4            17         1   4        0   \n",
       "23            24.7            13         0   4        0   \n",
       "24            24.7            13         0   3        0   \n",
       "25            27.4            17         1   3        0   \n",
       "26            27.4            17         1   4        0   \n",
       "27            27.4            17         1   3        0   \n",
       "28            27.4            17         1   2        0   \n",
       "29            27.4            17         1   4        0   \n",
       "30            24.7            13         0   4        0   \n",
       "31            24.7            13         0  -1        1   \n",
       "32            27.4            17         1   3        0   \n",
       "33            24.7            13         0   4        0   \n",
       "34            27.4            17         1   3        0   \n",
       "35            27.4            17         1  -1        3   \n",
       "36            27.4            17         1   3        0   \n",
       "37            27.4            17         1   3        0   \n",
       "38            27.4            17         1  -1        3   \n",
       "39            27.4            17         1   0        0   \n",
       "40            27.4            17         1   3        0   \n",
       "\n",
       "                                             spec_seq   annotation  \\\n",
       "0                               GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "1                                   GAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "2                                GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "3                               CAGAGTTCTACAGTCCGGCCT  isomiR-mRNA   \n",
       "4                                   GAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "5                               GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "6                                CAGAGTTCTACAGTCCGCCT  isomiR-mRNA   \n",
       "7                               GAGTTCTACAGTCCGACCAAT  isomiR-mRNA   \n",
       "8          GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "9                                   TTCAGAGTTCTACAGTC  isomiR-mRNA   \n",
       "10                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "11                              GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "12                                 GAGTTCTACAGTCCGACG  isomiR-mRNA   \n",
       "13                                CAGAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "14                                  GAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "15                                TTCAGAGTTCTACAGTCCC  isomiR-mRNA   \n",
       "16                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "17                             AGAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "18                             AGAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "19                                  TTCCGAGTTCTACAGTC  isomiR-mRNA   \n",
       "20                              GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "21                             CAGAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "22                              GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "23                                 TTCAGAGTTCTACAGTCC  isomiR-mRNA   \n",
       "24                               TCAGAGTTCTACAGTCCCCT  isomiR-mRNA   \n",
       "25                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "26                              GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "27                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "28                        CAGAGTTCTACAGTCCGACGAATGCCT  isomiR-mRNA   \n",
       "29                              GAGTTCTACAGTCCGACGATC  isomiR-mRNA   \n",
       "30                                  TTCAGAGTTCTACAGTC  isomiR-mRNA   \n",
       "31                                  GAGTTCTACAGTCCCCT  isomiR-mRNA   \n",
       "32                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "33  TTCAGAGTTCTACAGTCCTTTTTTTTTTTAATGATACGGCGACCAC...  isomiR-mRNA   \n",
       "34                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "35                                  GAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "36                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "37                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "38             GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "39                                  GAGTTCTACAGTCCGAC  isomiR-mRNA   \n",
       "40                               GAGTTCTACAGTCCGACGAT  isomiR-mRNA   \n",
       "\n",
       "            final_seq                                       chimeric_seq  \n",
       "0   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGCAACGTGTCAGGCACACAGACATA...  \n",
       "1   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAGCAGCAGGAGCAAGAGAAGCTGGA...  \n",
       "2   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAGAAGAAGGATGAGGGGGCTCAGCG...  \n",
       "3       GAGTTCTACAGTC  AGCTGGCAGAGAGGCTGATGACATCGTGAACTGGCTGAAGAAGCGC...  \n",
       "4   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCCGCGCGCGGGTCGGGGGGCGGGGCG...  \n",
       "5   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAAGGAAAAACTCATTGCACCAGTTG...  \n",
       "6       GAGTTCTACAGTC  GGTGGGCACCCAGGGACGGCATGCCGGGGGCCTGGGCAAGGAGTTC...  \n",
       "7   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACCAATTGGTGAAACCCCGTCTCTACTAAAA...  \n",
       "8   GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGACGATCTAAC...  \n",
       "9       GAGTTCTACAGTC  ACCGCGCGGCCCTGGCGGGCAGATGACATCCTGGCCAGCCCCCCTC...  \n",
       "10  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAGGCCGTCATGGAGACGCTGTTGCA...  \n",
       "11  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAAGGAACAAGATCCGAGCAATAGGC...  \n",
       "12  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGTCTACGGCCATACCACCCTGAACG...  \n",
       "13  GAGTTCTACAGTCCGAC  CAGAGTTCTACAGTCCGACGATCCAGATTGTGACATTTGCAGAGAG...  \n",
       "14  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCCTGGTGGATCGCCGACTGCAGGTGA...  \n",
       "15      GAGTTCTACAGTC  ACGGGCTGTGGGAAGACCACGCGCATCCCCCAGCTGTTGCTGGAGC...  \n",
       "16  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGTGATTTGGAGAGAGGTGTGGAAAT...  \n",
       "17  GAGTTCTACAGTCCGAC  AGAGTTCTACAGTCCGACGATCATGCGACCTCATCACCAGCACCAA...  \n",
       "18  GAGTTCTACAGTCCGAC  AGAGTTCTACAGTCCGACGATCAGGGGTAGATCTCGTTCTAGAACA...  \n",
       "19      GAGTTCTACAGTC  TTTGAATGCTACCCTGACCACAGTGCTTTGGACTTTCCGAGTTCTA...  \n",
       "20  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCACGGACCCAGCTCAGTGAGCTTCGC...  \n",
       "21  GAGTTCTACAGTCCGAC  CAGAGTTCTACAGTCCGACGATCAGGGCTGCGCCCAGAGTCCCAGG...  \n",
       "22  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCCTGATCTGCTTCCATCTCCAACTGG...  \n",
       "23      GAGTTCTACAGTC  TTGGGGCTGGTGGGACTTTCCTTCTGTGGGGCGGATGTGGGTGGCT...  \n",
       "24      GAGTTCTACAGTC  GTGCAGCGCAGCCTGGTGCTGCAGCGACCACAGACCTGGGCTGACT...  \n",
       "25  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAGCTGCTGGCCCCGGCCACCCAGCC...  \n",
       "26  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGCAACGTGTCAGGCACACAGACATA...  \n",
       "27  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAGGGCTGCGCCCAGAGTCCCAGGCC...  \n",
       "28  GAGTTCTACAGTCCGAC  AGCGAACACGATGATGAACGCGTTCAGAGTTCTACAGTCCGACGAA...  \n",
       "29  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCTGGATGGATTTGGACAGGAAATCTT...  \n",
       "30      GAGTTCTACAGTC  AAGCTCCACCTGGACTACATCGGGCCTTGCAAATACATCCCCCCTT...  \n",
       "31      GAGTTCTACAGTC  AGCGGCGCAGACTCGGGGGCAGGTTTTTGTGTAACGCTGCGGTCAC...  \n",
       "32  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAAACGGAGCTGAGCTGGCCTGGGTC...  \n",
       "33      GAGTTCTACAGTC  AAGTGGGGCGATGCTGGCGCTGAGTACGTCGTGGAGTCCACCGAGA...  \n",
       "34  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCACTCCGGATGTGCTGACCCCTGCGA...  \n",
       "35  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCCGCGCGCGGGTCGGGGGGCGGGGCG...  \n",
       "36  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCAGCCGGTCCCCCGCTGTCCCCGCCG...  \n",
       "37  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGAATGGATGAACGAGATTCCCACTG...  \n",
       "38  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCGAGTTCTACAGTCCGACGATCCGCG...  \n",
       "39  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCACGAAGATACATGTGTTTGACTACC...  \n",
       "40  GAGTTCTACAGTCCGAC  GAGTTCTACAGTCCGACGATCCGAGGCGTGGGTGGGCAGCCGTTCC...  \n",
       "\n",
       "[41 rows x 32 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isomir_df = mrna_mirna_df[mrna_mirna_df['annotation'] == 'isomiR-mRNA']\n",
    "isomir_df.reset_index(drop=True, inplace=True)\n",
    "isomir_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isomir_df.to_csv(r'C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\eclip\\SRR18281063\\R1\\isomir.csv', index= False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Categorizing isomiRNAs into different types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isomir_df = pd.read_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\eclip\\SRR18280895\\R1\\isomir.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\16892\\AppData\\Local\\Temp\\ipykernel_14908\\510302216.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isomir_df['isomir_type'] = \"\"\n"
     ]
    }
   ],
   "source": [
    "def check_mismatch_positon(chim_mirna, ori_mirna):\n",
    "    if len(chim_mirna) != len(ori_mirna):\n",
    "        raise ValueError(\"The length between chim_mirna and ori_mirna is not equal\")\n",
    "    \n",
    "    for i in range(len(chim_mirna)):\n",
    "        if chim_mirna[i] != ori_mirna[i]:\n",
    "            return 1\n",
    "    return 0\n",
    "\n",
    "def get_start_end_ind(chim_mirna, ori_mirna):\n",
    "    match = re.search(ori_mirna, chim_mirna)\n",
    "    if match:\n",
    "        start_index = match.start()\n",
    "        end_index = match.end() - 1\n",
    "        return start_index, end_index\n",
    "    else:\n",
    "        raise ValueError(\"chim_mirna does not have ori_mirna\")\n",
    "    \n",
    "\n",
    "isomir_df['isomir_type'] = \"\"\n",
    "\n",
    "for i in range(len(isomir_df)):\n",
    "    chim_mirna = isomir_df['final_seq'][i] # isomiRNA from chimeric sequence\n",
    "    ori_mirna = sequences[isomir_df['saccver_miRNA'][i]] #original miRNA sequence from miRBASE\n",
    "    ori_mirna_portion = isomir_df['sseq_miRNA'][i] #aligned miRNA portion from original miRNA sequence\n",
    "    mismatch = isomir_df['mismatch'][i]\n",
    "    front_distance = isomir_df['sstart_miRNA'][i] - 1\n",
    "    back_distance = isomir_df['slen_miRNA'][i] - isomir_df['send_miRNA'][i]\n",
    "    start_ind = isomir_df['sstart_miRNA'][i]\n",
    "    end_ind = isomir_df['send_miRNA'][i]\n",
    "    strand_info = isomir_df['saccver_miRNA'].iloc[i][-2:]\n",
    "    type = []\n",
    "\n",
    "    if len(chim_mirna) == len(ori_mirna):\n",
    "        if mismatch == 1:\n",
    "            type.append(\"SNP\")\n",
    "        else:\n",
    "            type.append(\"MNP\")\n",
    "\n",
    "    elif len(chim_mirna) > len(ori_mirna):\n",
    "        chim_start_ind, chim_end_ind = get_start_end_ind(chim_mirna, ori_mirna)\n",
    "        front_add = chim_start_ind - 0\n",
    "        back_add = len(chim_mirna) - chim_end_ind - 1\n",
    "\n",
    "        if front_add != 0:\n",
    "            if strand_info == '5p':\n",
    "                type.append(\"5p-addition\")\n",
    "            else:\n",
    "                type.append(\"3p-addition\")\n",
    "\n",
    "        if back_add != 0:\n",
    "            if strand_info == '5p':\n",
    "                type.append(\"3p-addition\")\n",
    "            else:\n",
    "                type.append(\"5p-addition\")\n",
    "\n",
    "        if (front_add > 4) & (back_add > 4):\n",
    "            new_seq = chim_mirna[chim_start_ind-4:chim_end_ind+4+1]\n",
    "            isomir_df.at[i, 'final_seq'] = new_seq\n",
    "        else:\n",
    "            if front_add > 4:\n",
    "                new_seq = chim_mirna[chim_start_ind-4:]\n",
    "                isomir_df.at[i, 'final_seq'] = new_seq\n",
    "\n",
    "            if back_add > 4:\n",
    "                new_seq = chim_mirna[:chim_end_ind+4+1]\n",
    "                isomir_df.at[i, 'final_seq'] = new_seq \n",
    "    else:\n",
    "        if start_ind != 1:\n",
    "            if strand_info == '5p':\n",
    "                type.append(\"5p-deletion\")\n",
    "            else:\n",
    "                type.append(\"3p-deletion\")\n",
    "        \n",
    "        if check_mismatch_positon(chim_mirna[0:4-start_ind+1], ori_mirna_portion[0:4-start_ind+1]) == 1:\n",
    "            if strand_info == '5p':\n",
    "                type.append('5p-replacement')\n",
    "            else:\n",
    "                type.append('3p-replacement')\n",
    "\n",
    "        if back_distance != 0:\n",
    "            if strand_info == '5p':\n",
    "                type.append('3p-deletion')\n",
    "            else:\n",
    "                type.append(\"5p-deletion\")\n",
    "\n",
    "        if check_mismatch_positon(chim_mirna[-(4-back_distance):], ori_mirna_portion[-(4-back_distance):]) == 1:\n",
    "            if strand_info == '5p':\n",
    "                type.append('3p-replacement')\n",
    "            else:\n",
    "                type.append('5p-replacement')\n",
    "            \n",
    "    isomir_df.at[i, 'isomir_type'] = type            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isomir_type\n",
       "[SNP]            31\n",
       "[5p-deletion]    10\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isomir_df['isomir_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isomir_df.to_csv(r\"C:\\Users\\16892\\OneDrive - University of Central Florida\\BIOINFORMATICS LAB\\isomir_dataset_paper\\outputs\\fields\\SRR14432306\\isomir_annotation.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
